src/pudl/workspace/setup.py

```diff
@@ -1,15 +1,114 @@
+            that value. Derived paths treat the parent directory as the input
+            workspace root.
+        output_dir: equivalent to PUDL_OUTPUT environment variable, but
+            overrides that value. Derived paths treat the parent directory as
+            the output workspace root.
+        yaml_file: a buffer including the YAML configuration. The `pudl_in` and
+            `pudl_out` keys within this file correspond to the input/output
+            workspace roots directly, instead of through parents.
+
+    Returns:
+        dictionary with a variety of different paths where inputs/outputs are
+        to be found.
+    """
+    load_dotenv()
+
+    # Workaround for not having PUDL_* env vars in ReadTheDocs builds.
+    #
+    # They don't let you set env var through config files, and I'd rather
+    # have this in source control than go through some sort of web UI
+    #
+    # I don't like this any more than you do.
+    if os.getenv("READTHEDOCS"):
+        os.environ["PUDL_OUTPUT"] = str(Path("~/pudl-work/output").expanduser())
+        os.environ["PUDL_INPUT"] = str(Path("~/pudl-work/data").expanduser())
+
+    # read from YAML source
+    if yaml_file is not None:
+        yaml_settings = yaml.safe_load(yaml_file)
+    elif default_pudl_yaml and default_pudl_yaml.exists():
+        with default_pudl_yaml.open() as f:
+            yaml_settings = yaml.safe_load(f)
+    else:
+        yaml_settings = {}
+
+    # read from env vars
+    env_var_mapping = {
+        "pudl_in": os.getenv("PUDL_INPUT"),
+        "pudl_out": os.getenv("PUDL_OUTPUT"),
+    }
+    env_settings = {
+        key: str(Path(value).parent)
+        for key, value in env_var_mapping.items()
+        if value is not None
+    }
+
+    # read from params
+    kwarg_mapping = {"pudl_in": input_dir, "pudl_out": output_dir}
+    kwarg_settings = {
+        key: str(Path(value).parent)
+        for key, value in kwarg_mapping.items()
+        if value is not None
+    }
+
+    # Start with an empty settings, then override in order of precedence.
+    settings: dict[str, str] = {}
+    for settings_source in [yaml_settings, env_settings, kwarg_settings]:
+        settings |= settings_source
+
+    if not ("pudl_in" in settings and "pudl_out" in settings):
+        raise RuntimeError(
+            "Must set 'PUDL_OUTPUT'/'PUDL_INPUT' environment variables or provide valid yaml config file."
+        )
+
+    settings = derive_paths(settings["pudl_in"], settings["pudl_out"])
+
+    # override env vars so Dagster can see the most up-to-date configs
+    if output_dir or "PUDL_OUTPUT" not in os.environ:
+        os.environ["PUDL_OUTPUT"] = settings["pudl_out"]
+    if input_dir or "PUDL_INPUT" not in os.environ:
+        os.environ["PUDL_INPUT"] = settings["data_dir"]
+    if "DAGSTER_HOME" not in os.environ:
+        os.environ["DAGSTER_HOME"] = str(Path(settings["pudl_in"]) / "dagster_home")
```

Comment:
If the `DAGSTER_HOME` env var isn't set prior to launching dagit, dagit will still use a temporary directory. If folks then restart dagit `DAGSTER_HOME` will then be set to `settings["pudl_in"]) / "dagster_home"`?

Reply:
This os.environ setting should only persist for this process & its child processes.

Reply:
We ended up deciding that if you're using dagit you just have to set the env vars, so this function won't be called on a normal dagit launch (we did this so we wouldn't have to call `get_defaults` directly from `pudl/etl/__init__.py` which would happen when anytime you `import pudl.etl`). Either way, I don't think changes to `environ` persist between runs.